From: Ben Hutchings <ben@decadent.org.uk>
Date: Sun, 8 Jan 2012 00:56:30 +0000
Subject: [PATCH 2/2] sparc: Change io_remap_pfn_range() back into an extern,
 exported function

In 3.1.8 io_remap_pfn_range() and remap_pfn_range() were reimplemented,
with the former becoming an inline function.  For ABI compatibility,
change it back into an extern function and export it.
---
 arch/sparc/include/asm/pgtable_64.h |   15 +++------------
 arch/sparc/mm/Makefile              |    2 +-
 arch/sparc/mm/generic_64.c          |   16 ++++++++++++++++
 3 files changed, 20 insertions(+), 13 deletions(-)
 create mode 100644 arch/sparc/mm/generic_64.c

diff --git a/arch/sparc/include/asm/pgtable_64.h b/arch/sparc/include/asm/pgtable_64.h
index 38ebb2c..fc36841 100644
--- a/arch/sparc/include/asm/pgtable_64.h
+++ b/arch/sparc/include/asm/pgtable_64.h
@@ -768,18 +768,9 @@ extern int page_in_phys_avail(unsigned long paddr);
 extern int remap_pfn_range(struct vm_area_struct *, unsigned long, unsigned long,
 			   unsigned long, pgprot_t);
 
-static inline int io_remap_pfn_range(struct vm_area_struct *vma,
-				     unsigned long from, unsigned long pfn,
-				     unsigned long size, pgprot_t prot)
-{
-	unsigned long offset = GET_PFN(pfn) << PAGE_SHIFT;
-	int space = GET_IOSPACE(pfn);
-	unsigned long phys_base;
-
-	phys_base = offset | (((unsigned long) space) << 32UL);
-
-	return remap_pfn_range(vma, from, phys_base >> PAGE_SHIFT, size, prot);
-}
+extern int io_remap_pfn_range(struct vm_area_struct *vma,
+			      unsigned long from, unsigned long pfn,
+			      unsigned long size, pgprot_t prot);
 
 #include <asm-generic/pgtable.h>
 
diff --git a/arch/sparc/mm/Makefile b/arch/sparc/mm/Makefile
index 301421c..87e825e 100644
--- a/arch/sparc/mm/Makefile
+++ b/arch/sparc/mm/Makefile
@@ -4,7 +4,7 @@
 asflags-y := -ansi
 ccflags-y := -Werror
 
-obj-$(CONFIG_SPARC64)   += ultra.o tlb.o tsb.o gup.o
+obj-$(CONFIG_SPARC64)   += ultra.o tlb.o tsb.o gup.o generic_64.o
 obj-y                   += fault_$(BITS).o
 obj-y                   += init_$(BITS).o
 obj-$(CONFIG_SPARC32)   += loadmmu.o
diff --git a/arch/sparc/mm/generic_64.c b/arch/sparc/mm/generic_64.c
new file mode 100644
index 0000000..d707319
--- /dev/null
+++ b/arch/sparc/mm/generic_64.c
@@ -0,0 +1,16 @@
+#include <linux/mm.h>
+#include <linux/module.h>
+
+int io_remap_pfn_range(struct vm_area_struct *vma,
+		       unsigned long from, unsigned long pfn,
+		       unsigned long size, pgprot_t prot)
+{
+	unsigned long offset = GET_PFN(pfn) << PAGE_SHIFT;
+	int space = GET_IOSPACE(pfn);
+	unsigned long phys_base;
+
+	phys_base = offset | (((unsigned long) space) << 32UL);
+
+	return remap_pfn_range(vma, from, phys_base >> PAGE_SHIFT, size, prot);
+}
+EXPORT_SYMBOL(io_remap_pfn_range);
-- 
1.7.8.2

